L1Trigger/L1THGCal/src/backend_emulator/HGCalStage1TruncationImpl_SA.cc

```diff
@@ -0,0 +1,108 @@
+#include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalStage1TruncationImpl_SA.h"
+#include <cmath>
+#include <iostream>
+
+using namespace l1thgcfirmware;
+
+HGCalStage1TruncationImplSA::HGCalStage1TruncationImplSA() {}
+
+unsigned HGCalStage1TruncationImplSA::run(const l1thgcfirmware::HGCalTriggerCellSACollection& tcs_in,
+                                          const l1thgcfirmware::Stage1TruncationConfig& theConf,
+                                          l1thgcfirmware::HGCalTriggerCellSACollection& tcs_out) const {
+  std::unordered_map<unsigned, std::vector<l1thgcfirmware::HGCalTriggerCell>> tcs_per_bin;
+
+  // configuation:
+  bool do_truncate = theConf.doTruncate();
+  const std::vector<unsigned>& maxtcsperbin = theConf.maxTcsPerBin();
+
+  // group TCs per (r/z, phi) bins
+  for (const auto& tc : tcs_in) {
+    unsigned roverzbin = tc.rOverZ();
+    int phibin = tc.phi();
+    if (phibin < 0)
+      return 1;
+    unsigned packed_bin = packBin(roverzbin, phibin);
+
+    tcs_per_bin[packed_bin].push_back(tc);
+  }
+
+  for (auto& bin_tcs : tcs_per_bin) {
+    unsigned roverzbin = 0;
+    unsigned phibin = 0;
+    unpackBin(bin_tcs.first, roverzbin, phibin);
+
+    const unsigned ntcin = std::max(smallerMultOfFourGreaterThan(bin_tcs.second.size()),
+                                    smallerMultOfFourGreaterThan(maxtcsperbin[roverzbin]));
+    const unsigned ntcout = (do_truncate ? maxtcsperbin[roverzbin] : bin_tcs.second.size());
+
+    l1thgcfirmware::HGCalStage1SortingAlg_SA tcSorter(ntcin, ntcout);
+
+    std::vector<unsigned> theTCsIn_mipt(ntcin);
+    std::vector<unsigned> theTCsOut_mipt(ntcout);
+    std::vector<unsigned> theTCsOut_addr(ntcout);
+
+    for (unsigned i = 0; i < ntcin; ++i) {
+      if (i < bin_tcs.second.size())
+        theTCsIn_mipt[i] = bin_tcs.second.at(i).energy();
+      else
+        theTCsIn_mipt[i] = 0;
+    }
+
+    tcSorter.sorting(theTCsIn_mipt, theTCsOut_mipt, theTCsOut_addr);
+
+    for (const unsigned& tcid : theTCsOut_addr) {
+      if (tcid < bin_tcs.second.size()) {
+        tcs_out.push_back(bin_tcs.second.at(tcid));
+      } else {  // Create and push a dummy TC with 0 energy and identifiable address
+        l1thgcfirmware::HGCalTriggerCell dummytc(false, false, roverzbin, phibin, 0, 0);
+        dummytc.setIndex(255);                         // set moduleID as done in firmware
+        dummytc.setCmsswIndex(std::make_pair(0, 63));  // set TC ID as done in firmware
```

Comment:
How CMSSW index could be related to what is done in firmware?
If there is no associated CMSSW TC it's not clear why `63` should be used. In principle there could exist a real CMSSW index `63`